packages/web/src/hooks/useCollapse.ts

```diff
@@ -0,0 +1,41 @@
+import { useState, useEffect, MutableRefObject } from 'react'
+
+import { Nullable } from 'common/utils/typeUtils'
+
+const TRUNCATE_LINE_HEIGHT = 16
+const MAX_NUM_LINES_AFTER_TRUNCATE_WEB = 4
+
+type UseCollapseProps = {
+  ref: MutableRefObject<Nullable<HTMLElement>>
+  lineHeight?: number
+  numLines?: number
+}
+
+export const useCollapse = ({
+  ref,
+  lineHeight = TRUNCATE_LINE_HEIGHT,
+  numLines = MAX_NUM_LINES_AFTER_TRUNCATE_WEB
+}: UseCollapseProps) => {
+  const [isCollapsible, setIsCollapsible] = useState(false)
+  const [isCollapsed, setIsCollapsed] = useState(false)
+
+  useEffect(() => {
+    if (ref?.current && !isCollapsed && !isCollapsible) {
+      const height = parseInt(
+        document.defaultView
```

Comment:
TIL! This is a new way of measuring element height for me - I'm more used to seeing getBoundingClientRect() or offsetHeight. Did some research though and it looks like `document.defaultView` is unnecessary? https://developer.mozilla.org/en-US/docs/Web/API/Window/getComputedStyle#defaultview